src/dso_hdr.cc

```diff
@@ -243,21 +243,36 @@ DDProfModRange DsoHdr::compute_mod_range(DsoMapConstIt it, const DsoMap &map) {
     return DDProfModRange();
   }
   DsoMapConstIt first_el = it;
-  while (first_el != map.begin()) {
-    --first_el;
-    if (it->second._filename != first_el->second._filename) {
-      ++first_el;
-      break;
+  {
+    DsoMapConstIt loop_el = first_el;
+    while (loop_el != map.begin()) {
+      --loop_el;
+      // there can be anonymous regions in the middle of the mapping
+      if (loop_el->second._type != dso::DsoType::kAnon &&
+          it->second._filename != loop_el->second._filename) {
+        break;
+      }
+      if (it->second._filename == loop_el->second._filename) {
+        first_el = loop_el;
+      }
     }
   }
-  while (++it != map.end()) {
-    if (it->second._filename != first_el->second._filename) {
-      break;
+
+  DsoMapConstIt last_el = it;
+  {
+    DsoMapConstIt loop_el = last_el;
+    while (++loop_el != map.end()) {
```

Comment:
I understand that we don't consume `loop_el` when `loop_el == map.end()`.  However, since the iterator doesn't have first-order knowledge of the container boundaries and `if (loop_el == map.end()) ++loop_el;` is UB, does this loop need to be structured similarly to lines 248-249?

Genuine question, I really have no idea.

Reply:
Thanks for checking loop structure
- End element is an invalid element after all elements. We should not process it.
- Begin element is valid and should be processed. Also we can not do a -- on the begin element.

The order of operators means that `++` is done prior to the evaluation of the `==`
If I had written on lines 248
```
    while (--loop_el != map.begin()) {
//...
```
I would not have processed the begin element. Alternatively I think a `do...while` would have been better.... Feel free to tell me if you have a more stylish way of writing this

Reply:
Ah, rookie mistake on my part--thanks for clarifying.